### DISCUSSION TITLE
Best way to publish individual components of a UI Library to npm?

### DISCUSSION THREAD

**Discussion by theolavaux** [vercel/turborepo#3754] [Help] [answered]
I'm creating a monorepo with several apps that have to share a common UI Library.
To avoid breaking changes in all apps when a component is modified, I want to introduce a versionning system and publish individually each component of my library to a private npm registry (e.g. `@my-company/ui-button`, `@my-company/ui-select`, etc.)

However, a few questions arise on the best way to achieve this :

1. Should I keep my whole component library as a Turborepo package or add individual components to the `packages`?

2. What could be the dev flow of developing and publishing a package with the CI and then use it in an app with its specific version? 

3. I would like to create a progress to test, build and publish the components that have changed to a private NPM registry. The more automated it could be, the best it would be according to me. How can I detect which folders (=components) have changed inside my UI library and run a specific set of scripts for them? I thought about using Lerna in my components library to publish each component but it might conflict/complicate things with Turborepo to have nested packages inside a package.

5. How to keep a clear changelog for each component to know what has been changed in each version?

If anyone has been through this before or has resources that can guide me tot he right direction, it would be highly appreciated!

If you think that publishing the library as a whole is better, let me know how you handle dependencies (if two packages have breaking changes for instance but you only want to bump one package's version).

[Here is the template that I used](https://github.com/gurvancampion/turborepo-nuxt-boilerplate) (and modified a bit) to start the project

**Comment by anthonyshew** [comment, marked as answer]
There are a few architectural issues to what is being described here. This describes _a lot_ of overhead that is generally antithetical to how monorepos work (unless you're meaning you want to publish for external consumers).

If you are indeed publishing for external consumers, your best bet is to make one package and publish it. Changelogging can be handled a variety of ways. Here's an example of how we handle the changelog for `turbo` (use github labels to generate heading in copy): https://github.com/vercel/turbo/blob/main/.github/turborepo-release.yml

That said, this is a lot of custom setup that I don't think we would be able to give general recommendations on, both in and out of scope of Turborepo.
